Add documentation (#989)
MetaMask/controllers: docs/what.md

# All about controllers

If you've come here from the [README](../README.md), you already know that this repo holds controllers. But what even _is_ a controller?

The short answer is that it's our solution to managing and disseminating state changes within our products.

> Every application needs to manage state. Some state is temporary. Perhaps the

Reviewer: Hmm, this paragraph seems a bit... meandering. I like the idea of establishing what application state is; just saying "state" can be a bit too abstract to grasp. Maybe we can reformat this to separate the examples of application state from the details about some state being transient, some persisted, etc.

> Every application needs to manage state. Some state is temporary. Perhaps the
> user is on a screen which keeps track of whether a button is clickable/tappable depending on whether a request is taking place; but as soon as the user leaves that screen, that information is discarded. But some state is more permanent. Perhaps a screen needs to reuse information that was gathered on a previous screen, in which case the state is cached in memory; or perhaps the entire application needs to load all of the user's data when it starts, in which case the state is persisted to disk. Either way, we need a place to store that state, and that's where controllers come into play.

Reviewer 2: Unexpected
linebreak
```suggestion
Every application needs to manage state. Some state is temporary. Perhaps the user is on a screen which keeps track of whether a button is clickable/tappable depending on whether a request is taking place; but as soon as the user leaves that screen, that information is discarded. But some state is more permanent. Perhaps a screen needs to reuse information that was gathered on a previous screen, in which case the state is cached in memory; or perhaps the entire application needs to load all of the user's data when it starts, in which case the state is persisted to disk. Either way, we need a place to store that state, and that's where controllers come into play.
```

Author: Haha. Just an editor thing. :)

> Currently, there are two styles of creating controllers which we've dubbed [BaseController v1](../packages/base-controller/src/BaseController.ts) and [BaseController v2](../packages/base-controller/src/BaseControllerV2.ts). Both versions offer the same set of basic functionality:

Reviewer 2: I know it's already prevalent but IMO we (see what I did there) should avoid speaking from  first-person in docs like this - reserve that for discussion and P2P comms.

```suggestion
There are currently two styles of creating controllers, each with an associated TypeScript base class: [BaseController v1](../packages/base-controller/src/BaseController.ts) and [BaseController v2](../packages/base-controller/src/BaseControllerV2.ts). Both versions offer the same set of basic functionality:
```

Author: I can probably guess the reasoning for this, so I don't disagree, but I'm just curious — why in your opinion should this be avoided?

Reviewer 2: There is more to it but here is one aspect: By introducing the "we", the reader is now forced to consider the author explicitly; who is writing this, when? In case of some open-source projects: Is this a message from the original author, the intermediary repository maintainers, or the company now owning the codebase? Regardless of if the reader thinks this matters or not, the context is now more dynamic. Technical documentation relates to the subject, independently of people and organizations.

For example, if we consider an alternative reality where we're documenting a proprietary hosted HTTP API, documenting rate-limiting behavior - this is one example of when I think a "we" could be warranted, depending on style.

Does that make sense?

Author: Yeah! I hadn't thought about that but it makes perfect sense.

> - The developer can preload the state (say, from a persistent location) when initializing.
> - The application can update the state at any time.
> - In another location, the application can listen for and respond to state changes (or stop listening altogether).

Reviewer: Not sure about the use of "location" here. The first time it seems like it means a location on disk, and the second it seems like you mean somewhere else in the codebase.

Maybe something like this would be clearer:

```suggestion
- State can be set as part of initialization.
- State can be updated at any time.
- Controllers allow subscribing to state updates.
```

Author: Ah, gotcha, good point. 

I was trying to make it a bit more visualizable rather than abstract, and I feel like using the active form achieves that more than passive, but your rewording is more concise, so I like that. I'll workshop this.

> v2 includes a number of improvements over v1, resulting in a more simplified and solidified API. One aspect worth noting with v2, however, is that not only can controllers communicate with the application, but controllers can also communicate with each other — and this is done in a way that doesn't require the controller package that's doing the receiving to depend on the controller that's doing the sending. This capability is provided by [ControllerMessenger](../packages/base-controller/src/ControllerMessenger.ts).

Reviewer 2: I'm not sure I got it right here but somehow I think that part can be simplified a bit?

```suggestion
v2 includes a number of improvements over v1, resulting in a more simplified and solidified API. One aspect worth noting with v2, however, is that not only can controllers communicate with the application, but controllers can also expose APIs for any other controllers to interface with. This capability is provided by [ControllerMessenger](../packages/base-controller/src/ControllerMessenger.ts).
```
